MCH: disable exception when digit array sizes don't match

aferrero2707 opened the pull request:
The sizes of the input and output digits arrays  do not always match in the case of continuous data, because there could be multiple digits for the same pad in the input buffer.

This will be fixed by the time-based pre-clustering that in future will run before the position-based pre-clustering.

aphecetche commented:
@aferrero2707 I'd let @pillot to comment as well, but instead of commenting the code, I would prefer a device option to be added to enable/disable this check instead ?

pillot commented:
I agree. Alternatively you can turn it into an error message for the moment, so that we can still see when this problem occurs.

In the end, even with the time-clustering (unless it is specifically protected against that), I suspect there might still be some peculiar (rare?) cases where this problem can happen, which is likely something we will want to catch.

aferrero2707 commented:
@aphecetche @pillot ok, I then propose to change the throw into a simple error message (I think we do not want to crash the whole DPL workflow when the problem appears), and allow disabling the check via a command-line switch.

Do you agree?

aphecetche commented:
I agree that we should avoid crashing workflows needlessly. But at the same time a simple error message has a high chance of just being ignored.
One solution that would allow the most flexibility is to have the new workflow option to be non boolean, e.g. something like `--check-no-leftover-digits=off | error (default) | fatal` which would respectively disable, make it an simple error message, or make it a fatal (crash) error. That last option can be used e.g. in tests.

aferrero2707 commented:
@aphecetche @pillot I have added an option as suggested.

aphecetche commented on Detectors/MUON/MCH/Workflow/src/PreClusterFinderSpec.cxx:
```diff
@@ -102,9 +117,19 @@ class PreClusterFinderTask
-      throw runtime_error("some digits have been lost during the preclustering");
-    }
+
+    // check sizes of input and output digits vectors
```
hum, where's the actual check now ? (for the cases the check-no-leftover-digits option is not off)

aferrero2707 replied:
Ooops! I should not do coding on Friday afternoon ;-)

I have added the missing check, and combined all the commits into a single one for clarity.

aferrero2707 committed:
[MCH]added option to enable/disable the check of the digits vectors sizes

The sizes of the input and output digits arrays  do not always match
in the case of continuous data, because there could be multiple
digits for the same pad in the input buffer.

This will be fixed by the time-based pre-clustering that will
run before the position-based pre-clustering.

aferrero2707 committed:
[MCH] fixed clang formatting

aferrero2707 committed:
[MCH] changed length of help message

aferrero2707 committed:
[MCH] using variable to store help string

aferrero2707 committed:
[MCH] fixed clang format
